### PR TITLE
Fix Gazebo freezing if YARP_CLOCK is set

### PR DESCRIPTION
**Pull request by traversaro**
Fix https://github.com/robotology/gazebo-yarp-plugins/issues/526 . 

TL;DR: From gazebo-yarp-plugins 3.6.0, you can again set YARP_CLOCK for the Gazebo simulation, meaning that all the threads of YARP Network Wrapper Servers will be executed with the frequency correctly synchronized with the Gazebo simulation. This also means that if you like you can also set `YARP_CLOCK`  directly in your `.bashrc` or `setup.sh` script. 

With respect to the solution discussed in https://github.com/robotology/gazebo-yarp-plugins/issues/526, i.e. : 
 
> 
> A possible solution is to switch to use the system clock after yarp::os::Network was initialized, and switch back to use the default clock once `/clock` has been created by calling `yarp::os::NetworkBase::yarpClockInit(YARP_CLOCK_DEFAULT)`. This should work fine _unless_ another Gazebo plugin called one of those methods to perturb the process-global YARP clock, but I think that is ok.

in this PR there are two modifications. 

First of all, the delayed initialization of the Clock is only done if the yarp::os::Network is not initialized, and this is checked by the `yarp::os::NetworkBase::isNetworkInitialized()` method. Secondly, calling `yarp::os::Network::useNetworkClock` does not only need the `/clock` port to exists, but also wait until something is published on that port and the NetworkClock becomes valid (see useNetworkClock). For this reason, calling `useNetworkClock` in the Gazebo Physics thread is going to deadlock, as the Gazebo physics thread is also the one that is publish the new time on the clock. To avoid this, the `yarp::os::Network::useNetworkClock`  is called as part of new `std::thread` that is just created for this.

The PR also contains a regression test for the issue, that was failing before the fix. As the issue was quite related to the inner working of YARP, it was not use `NetworkBase::setLocalMode` as in the robotinterface and controlboard tests, but rather yarpserver is actually launched before the tests by using [`tiny-process-library`](https://gitlab.com/eidheim/tiny-process-library).

### CODE REVIEW

File: plugins/clock/src/Clock.cc [hunk 1]
```diff
@@ -68,7 +68,20 @@ namespace gazebo
 
     void GazeboYarpClock::Load(int _argc, char **_argv)
     {
-        m_network = new yarp::os::Network();
+        // To avoid deadlock during initialation if YARP_CLOCK is set,
```
Comment (xEnVrE) [inline plugins/clock/src/Clock.cc:71]: Suggested change:
initialation -> initialization

**Review by xEnVrE — APPROVED** [review]
LGTM :+1: 

I only added few comments regarding a typo and a potential missing sentence in the code comments

File: plugins/clock/src/Clock.cc (hunk 1 above)
Comment (traversaro) [reply plugins/clock/src/Clock.cc:71]: Thanks, feel free to do this suggestions with GitHub's integrated system!

Comment (traversaro) [inline plugins/clock/src/Clock.cc:71]:
```suggestion
        // To avoid deadlock during initializationif YARP_CLOCK is set,
```
Like this!

File: plugins/clock/src/Clock.cc [hunk 2]
```diff
@@ -68,7 +68,20 @@ namespace gazebo
...
+        // if the YARP network is not initialized we always initialize
+        // it with system clock, and then we
```
Comment (traversaro) [inline plugins/clock/src/Clock.cc:73]:
```suggestion
        // it with system clock, and then we switch back to the default clock later
```

**Comment by traversaro** [comment]
> 
> 
> LGTM 👍
> 
> I only added few comments regarding a typo and a potential missing sentence in the code comments

I think I have addressed them. The tests are failing for the same reason mentioned in https://github.com/robotology/gazebo-yarp-plugins/issues/536, so as the tests will be suppressed in https://github.com/robotology/gazebo-yarp-plugins/pull/532 I think we can merge if you agree.

**Comment by xEnVrE** [comment]
>  I think we can merge if you agree.

Fine with me!
